### ISSUE TITLE
Using IAM roles for Amazon EC2 instances

### ISSUE THREAD

**Issue by jarrettj** [aws/aws-sdk-php#1054]
Hi, good day.

My EC2 instance has an IAM Role attached. The Role has full S3 access. Using the SDK I get the following error:

`Error retrieving credentials from the instance profile metadata server. (cURL error 28: Operation timed out after 1000 milliseconds with 0 bytes received (see http://curl.haxx.se/libcurl/c/libcurl-errors.html))`

The code is as follows:

```
$config = array(
         'region'   => 'us-east-1',
         'version'  => 'latest'
        );
$sdk = new Aws($config);
$s3Client = $sdk->createS3();
$file = $s3Client->getObjectUrl('mybucket', 'my/file');
```

If I add the credentials to the $config array it works. But I assumed the SDK would use the servers Role if no credentials were passed. Could you let me know if this is the behaviour or I am doing something wrong?

Thanks.

Regards.
JJ

**Comment by cjyclaire** [comment]
@jarrettj It's correct that SDK would use the servers Role if no credentials were passed in, no env variable is set and `~/.aws/credentials` is set. Could you double check that there is credentials attached with your instance correctly?

To see the key associated with your instance, you could execute this commandLine
```
curl http://169.254.169.254/latest/meta-data/iam/security-credentials/<role name>
```
To get the role name, you can run the following:
```
curl http://169.254.169.254/latest/meta-data/iam/security-credentials/
```

**Comment by jarrettj** [comment]
There's no ENV set, no ~/.aws/credentials file either. 

Doing:
`curl http://169.254.169.254/latest/meta-data/iam/security-credentials/`
gives nothing.

Maybe that is my issue? But the instance has a Role attached for sure! Maybe I should raise an issue with AWS. My test and live server returning nothing for that curl request.

When running:
`curl http://169.254.169.254/latest/meta-data/iam/security-credentials/aws-elasticbeanstalk-ec2-role`
it works though. And returns a success message with access key, secret and token in it.

The AWS CLI works without credentials as well. I did the following:
`aws s3 ls s3::/bucket/my/file`
That works as well.

Just the PHP SDK that flakes.

**Comment by jarrettj** [comment]
Solved.

I created a cloned environment from our production environment. The testing site however needed it's own test bucket for images. The images location bucket changed so obviously they do not exist in the test bucket. 

When doing a single call it works. But my application does a few hundred requests to images that do not exist and maybe that is why the credentials server stops responding.

A workaround, and good practice, is to cache the credentials. This avoids hitting the metadata service a couple of hundred times per request.

```
[
         'region'   => 'us-east-1',
         'version'  => 'latest',
         'bucket' => 'test-bucket',
         'credentials' => new \Aws\DoctrineCacheAdapter(new \Doctrine\Common\Cache\ApcuCache)
]
```

That worked for me. :)
